inst/rmarkdown/templates/template_spp_en/skeleton/08_subprotocols.Rmd

```diff
@@ -0,0 +1,20 @@
+# (PART\*) Subprotocols {-} 
+
+```{r, results="asis"}
+if (exists("dependencies")) {
+  mdfiles <- paste0(dependencies$protocol_code,
+                    "-", 
+                    dependencies$version_number, 
+                    ".md")
```

Comment:
Where does `add_subprotocols()` run? From its documentation, it should be run from a separate script. How does it fit in the workflow, where the dependencies & params are defined in the main `index.Rmd`? Should that script be present in the template, and sourced by the setup-chunk (for example)?

Reply:
The initial idea was to run it in an r chunk in that chapter. I tried that initially but then decided to run it in a separate r-script. This made it a little bit easier to debug. I am not even sure that, after debugging, `add_subprotocols()` would work if we move it back to the r-chunk. I think it's just not a good idea because the function calls `render_book()` and this call would be executed as part of the `knit` step when rendering the main protocol (so we get a `knit()` call in another `knit()` call and potentially lot's of issues with environments, locked bindings, ....).

Reply:
> so we get a knit() call in another knit() call

Sure, maybe won't work. But if it does, then that's probably the way to go. Somehow it must be compiled automatically, preferably in the same way for all protocols. Cf. the automated steps on release. How accomplish that?

Reply:
Since `add_subprotocols()` effectively writes the necessary md-files and associated files to subfolders of the main protocol, which are  then called from:

````
# (PART\*) Subprotocols {-} 

```{r, results="asis"}
if (exists("dependencies")) {
  mdfiles <- paste0(dependencies$protocol_code,
                    "-", 
                    dependencies$version_number, 
                    ".md")
  child_docs <- file.path(dependencies$version_number,
                          mdfiles)
  
  child_docs <- child_docs[dependencies$appendix]
  
  if (length(child_docs) > 0) {
    res <- map(child_docs, knit_child, quiet = TRUE)
    cat(unlist(res), sep = "\n")
  }
}
```
````

I don't see a problem with automatic compiling. You just have to make sure that during the PR for a new or updated spp the params in the `index.Rmd` are set correctly (so the object `dependencies` has the correct content) and that md-files of subprotocols have been added with `add_subprotocols()` (this needs to be done just once).

Reply:
> Where does `add_subprotocols()` run? From its documentation, it should be run from a separate script. How does it fit in the workflow, where the dependencies & params are defined in the main `index.Rmd`? Should that script be present in the template, and sourced by the setup-chunk (for example)?

No, it should not be sourced (see previous comment). But it is a good idea to add a script to the template folders that shows how to `add_subprotocols()`.

Reply:
> that md-files of subprotocols have been added with `add_subprotocols()` (this needs to be done just once).

I think I understand it now. I was a bit misguided by "_The function should be called in a separate R script_" in the documentation - just running it interactively (in the console) would then suffice. Maybe that should be stated more clear.

"to be done just once": conditional that nothing changes in the `dependencies` param, otherwise it needs to be run again IIUC. But indeed rendering would fail using the above code if one forgot to do that.

Some duplication here (from line 154 on):

https://github.com/inbo/protocolhelper/blob/d8ed68746261b73d273a4512db3ab911bc3ad238/tests/testthat/test-add_subprotocols.R#L143-L158

I.e. `add_subprotocols()` takes no advantage of the dependencies in `index.Rmd`. That may be worth considering. I.e. filling dependencies YAML block and then calling `add_subprotocols()` which already knows where to look e.g. just referring to code & version number of the main protocol.

Before, as `add_subprotocols()` has the expliciet dependencies arguments, I had expected it would use them while being called from _within_ the main bookdown project (i.e. as a first step in the chunk [above](https://github.com/inbo/protocolhelper/pull/42#discussion_r591551375)). So that does not work? _If_ it would, then it is the simplest workflow.

Reply:
I tried calling it within the main bookdown project, but it did not work (feel free to give it a try, I gave up debugging and believe it's just not possible). That is also the reason why I added lines 143-158 to test-add_subprotocols.R (so the correct `dependencies` object would be passed to `add_subprotocols()` in a chunk). I did not remove these lines, because it is good practice to set the params to the right values in the main protocol `index.Rmd` anyway.

Reply:
> I did not remove these lines, because it is good practice to set the params to the right values in the main protocol index.Rmd anyway

Sure!

What I meant is the repetition of the same values in lines 154-158: perhaps `add_subprotocols()` could be extended to read the dependencies from the YAML block in `index.Rmd` if dependency arguments are not given. It is not relevant for this testing file indeed, but such ability would ease the user's workflow in the console (not having to pass dependency values when running `add_protocols()`, but passing code & version number of the main protocol instead).

Reply:
Good suggestion. `add_subprotocols()` now only allows reading the dependencies from the yaml header. See f844529. This seems like the safest approach to me.